Adapt Superflare to work with Wrangler 3 and Remix v2 (vite compiler)

acusti opened the pull request:
This PR:

* updates the superflare/remix package + remix template + docs site to work with remix v2 and the vite compiler (based on the official [cloudflare workers template](https://github.com/remix-run/remix/tree/main/templates/cloudflare-workers))
* updates all packages to the latest v5.x version of typescript and fixes some new typescript errors
* updates the superflare CLI to work with wrangler v3 commands and miniflare v3 APIs. this entailed changes to some of the CLI command APIs:
    * the `migrate` command’s `--db` option is now the name of the DB binding, not a path to the actual local DB file
    * the `dev` command now runs `remix vite:dev` for the main server + `wrangler dev` in parallel to [enable Durable Objects](https://developers.cloudflare.com/workers/wrangler/api/#supported-bindings)
* removes direct dependency on `better-sqlite3` in favor of `wrangler`’s [`getPlatformProxy` API](https://developers.cloudflare.com/workers/wrangler/api/#getplatformproxy)
* removes `./d1js/*` vendored files
* adds a new export to the `@superflare/remix` plugin, available at `@superflare/remix/dev`, which provides `superflareDevProxyVitePlugin`, a drop-in replacement for [remix’s `cloudflareDevProxyVitePlugin`](https://remix.run/docs/en/main/guides/vite#cloudflare-proxy) that supports superflare and its automatic session handling and auth features
* upgrades to latest `pnpm`
* upgrades to latest `turbo`
* adopts new [Static Assets for Workers](https://developers.cloudflare.com/workers/static-assets/) feature

implements #62

acusti commented on apps/site/app/docs.server.ts:
```diff
@@ -88,14 +88,17 @@ export async function getManifest(
 
-function getNodeText(node: RenderableTreeNode) {
+function getNodeText(renderableNode: RenderableTreeNode) {
+  if (typeof renderableNode === "string") return renderableNode;
```
@jplhomer heads up that this line is a change in behavior. in the previous version of this function, if `typeof node === "string"`, the function returned an empty string, which i thought might be a bug, but if it is in fact desired, i will just lump this conditional in with the next when (line 93),

acusti commented:
@jplhomer i refactored the `migrate` command and the underlying utilities for accessing D1 database resources to work with wrangler v3 (and miniflare v3_, and i updated the remix template and @superflare/remix to work with remix v2 and the vite compiler. i’ve been able to test some of the CLI commands locally by building `packages/superflare` and then running `npx superflare …` from `examples/remix-cms`, but i’m not sure the best way to test things like `npx superflare new` (which seems to still use the last published version of `templates/remix`), and then cd into the created project and test the versions of other CLI commands from this branch.

do you have any suggestions on how to do that? also, please feel free to start reviewing or pull it down and check stuff out. the lines changed count is intimidating, but the vast majority of that is `pnpm-lock.yaml`:
<img width="482" alt="image" src="https://github.com/user-attachments/assets/80eaa319-4e9f-4a1c-9be8-51e83b36488b">

jplhomer commented on apps/site/worker.ts:
```diff
@@ -1,66 +1,56 @@
-    } catch (reason) {
-      console.error(reason);
-      return new Response("Internal Server Error", { status: 500 });
+      const loadContext = {
+        cloudflare: {
+          // This object matches the return value from Wrangler's
+          // `getPlatformProxy` used during development via Remix's
+          // `cloudflareDevProxyVitePlugin`:
+          // https://developers.cloudflare.com/workers/wrangler/api/#getplatformproxy
+          cf: request.cf,
+          ctx: { waitUntil, passThroughOnException },
+          caches,
+          env,
+        },
+      };
```
I think we'll need to package this up somehow in an exported function to be reused inside `vite.config` for development. E.g. when running the dev script right now, it complains about lacking `auth` and `session` indicating that they're not being injected into the context properly

acusti replied:
this logic is actually still packaged up in the `handleFetch` util in @superflare/remix, as here: https://github.com/jplhomer/superflare/pull/66/files#diff-c48c60214196d8cfdc34b781f4a782d02027358ef29712947d121ac4b010bac9R40

but this is the worker.ts for the docs site, so it doesn’t have superflare

jplhomer commented:
^ @acusti let's see if that helps us with the testing the scaffold/new flow! We may have to add an option to `new` for `--template` with a branch specifier or raw URL... we'll see.

jplhomer commented on package.json:
```diff
@@ -12,7 +12,8 @@
     "lint": "turbo run lint",
-    "format": "prettier --write \"**/*.{ts,tsx,md}\""
+    "format": "prettier --write \"**/*.{ts,tsx,md}\"",
+    "superflare": "./packages/superflare/bin/superflare.js"
```
FYI added this so you can `pnpm superflare <command>` more easily without the finicky behavior of `npx`. Also: `./scratch` exists as a locally ignored subfolder for purposes of scaffolding 👍

dan-gamble commented:
Just had a little play around with this, one thing that seems to need to be done now is within the `vite.config.ts` you need to pass the `AppLoadContext` too like:

```typescript
		cloudflareDevProxyVitePlugin({
			// @ts-ignore
			getLoadContext,
		})
```

I pulled the PR down and the `examples/remix-cms` doesn't have the `auth` and `session` on it when doing `pnpm dev`. It just has `cloudflare`. The `vite.config.ts` doesn't run through the `worker.ts` at all as far as I'm aware.

jplhomer commented:
@dan-gamble Yeah, one of the missing pieces. I think packaging it up as an exported function would help, or just adding a Superflare Vite plugin https://github.com/jplhomer/superflare/pull/66#discussion_r1693975128

acusti commented:
@jplhomer this is great, thanks so much for getting the `new` command working, and for the pkg.pr.new identifiers, as well as to you and to @dan-gamble for calling out the issues with injecting `auth` and `session` in dev.  i’m currently on the tail end of a week-long camping trip to kings canyon/sequoia national parks, but i’m excited to jump back into this monday morning.

dan-gamble commented:
I hope you're having a great time @acusti!

dan-gamble commented:
I also think the Vite introduction causes issues with things like queues & scheduled not working as the `worker/server.ts` never gets hit during dev. Only way I can think of getting around this, which is awful, is using a service binding to a standalone worker

acusti commented:
@jplhomer @dan-gamble i added a `getLoadContext` function to @superflare/remix to augment the load context with `auth` and `session` based on the pattern recommended in the [remix docs](https://remix.run/docs/en/main/guides/vite#augmenting-load-context) that you both alluded to already: fb38d74a7ccb5e844631441aeb0f2a22619d2e78

then i added a super simple `superflareDevProxyVitePlugin` convenience function (naming based on the `cloudflareDevProxyVitePlugin` exported from `@remix-run/dev`) that provides the new `getLoadContext` function to the vite plugin so that `auth` and `session` will now be available when running `superflare dev`: b9ecf0d2e071302929dfc3ace001ed5bf00cb1bc

makes for a nice clean diff when switching from plain remix + cloudflare workers to superflare: 031ab99e0410174b095bc9ae66cff474c632231c

right now i’m not exporting the custom `getLoadContext` function that i added to @superflare/remix because `handleFetch` and `superflareDevProxyVitePlugin` are already using it internally, but maybe that would be a useful thing to export.

also, i removed the local `load-context.ts` files from the remix template + remix-cms example, but i’m not sure how typescript deals with those ambient types and if this coming from the @superflare/remix package will still do the trick of updating @remix-run/cloudflare’s `AppLoadContext` type as it should.

dan-gamble commented:
That's super nice Andrew and makes a lot of sense. Did you have any thoughts around the `queue` and `scheduled` part not working with Vite?

Also, is there anything that I can help with in regards to this? I've been wanting to see this move for a _long_ time and would love to contribute in any kind of meaningful way!

dan-gamble commented:
My other though was around introducing ALS too. I was able to do this in a hacked together version of the old 2.x version of this. Something along the lines of this in the `worker.ts`

```typescript
    config({ env, request, ctx })

    const context = getContextFromUserConfig(config({ request, env, ctx }), env)
    context.env = env

    try {
      return await runWithContext(
        context,
        async () => handleRemixRequest(request, setupLoadContext(env, ctx)),
      )
    } catch (error) {
      console.log(error)
      return new Response('An unexpected error occurred', { status: 500 })
    }
```

However, with Vite, I'm not entirely sure how this will be feasible as well

jplhomer commented:
@acusti This is awesome! I'll take a look when I can.

@dan-gamble regarding queue consumers and scheduled workers: Queues is a bummer, yeah. Sounds like it's still not supported even in Pages dev (which this most closely resembled) https://developers.cloudflare.com/pages/functions/bindings/#interact-with-your-queue-producer-binding-locally

I'm less concerned about scheduled workers in local dev. In my experience, I rarely need to be running a scheduled job, unless I want to try a one-off version of it to on make sure it works. 

But it's still unfortunate we can't just migrate to Pages because of the lack of support for `scheduled` and `queue` in the Functions 😢 

Regardless, based on CF recommends (using a separate worker), I wonder if you're onto something with a service binding to something else, which we could then wire up for local dev purposes as well as provide some "glue" to deploy them in a nice way. 

Also: I got ALS mostly working https://github.com/jplhomer/superflare/pull/40 using a Vite plugin wrapper that binds the context to `globalThis`, similar to what you showed in your snippet. It'll be a bigger migration because of the impacts it has on things like testing, etc - so probably worth exploring in a separate workstream.

acusti commented:
@dan-gamble @jplhomer i made an update to use the new `getLoadContext` util to also instantiate the `config` singleton so that it can provide bindings as needed. i did that because `getLoadContext` is the common function invoked when handling requests across all environments (dev with vite and prod with workers), but i’m not sure if it’s the right solution.

i also realized that the implementation of `getLoadContext` that i pushed up yesterday wasn’t actually getting any of the bindings that matter because it wasn’t reading the `wrangler.json` file, so i added the `experimentalJsonConfig` option. that fixes it so that `getPlatformProxy` is actually reading the wrangler.json file, but it also means that it now errors on the `Channel` durable object, because the existing implementation relies on the user exporting that class from `worker.ts`, which is no longer being run.

that brought me back to the docs, where i found a [very helpful note](https://developers.cloudflare.com/workers/wrangler/api/#supported-bindings) about making Durable Objects work with `getPlatformProxy`:

To use a Durable Object binding with `getPlatformProxy`, always [specify a `script_name`](https://developers.cloudflare.com/workers/wrangler/configuration/#durable-objects) and have the target Worker run in a separate terminal via [`wrangler dev`](https://developers.cloudflare.com/workers/wrangler/commands/#dev).

For example, you might have the following file read by `getPlatformProxy`.

```toml
[[durable_objects.bindings]]
name = "MyDurableObject"
class_name = "MyDurableObject"
script_name = "my-worker"
```

In order for this binding to be successfully proxied by `getPlatformProxy`, a worker named `my-worker` with a Durable Object declaration using the same `class_name` of `"MyDurableObject"` must be run separately via `wrangler dev`.

-------------

so it seems we just need to run `wrangler dev` simultaneously as `remix vite:dev`, which seems ok to me? and like it could also mean that the `queue` and `scheduled` handlers will now get invoked? not sure yet if it will play well with the vite server, but i’m optimistic.

@dan-gamble i would love your help with this! one thing is that even putting aside the pnpm lock file, there are a lot of lines of changes, so i’m not sure how @jplhomer would like to handle code review. but i’m also very happy if you just want to jump in and make improvements or start trying to fix issues you come across in your own testing. we can use this PR to mention when we’re working on something and coordinate, or i’m happy to do a slack/discord kinda thing if that appeals, or if we wanna use GitHub issues/projects.

acusti commented:
quick update: i’m able to get the Config singleton defined with the bindings from `getPlatformProxy` + whatever is defined in `wrangler.json` (+ `superflare.config.ts`, of course). i’m currently testing that functionality by using the /auth/register form in the remix-cms, but for some reason, when `Model.getConnection` is called, the `Config` singleton no longer has any of the bindings. i’ve been logging out the `Config` from the end of `setConfig` and from within `Model.getConnection`, and here’s what those logs look like:

```js
// end of setConfig
[class Config] {
  listeners: Map(1) {
    'ArticleUpdated' => [ [class LogArticleUpdated extends Listener] ]
  },
  env: {
    APP_KEY: 'redacted',
    DB: ProxyStub { name: 'D1Database', poisoned: false },
    CHANNELS: ProxyStub { name: 'DurableObjectNamespace', poisoned: false },
    QUEUE: ProxyStub { name: 'WorkerQueue', poisoned: false },
    REMIX_CMS_MEDIA: ProxyStub { name: 'R2Bucket', poisoned: false }
  },
  appKey: 'redacted',
  database: {
    connections: { default: ProxyStub { name: 'D1Database', poisoned: false } }
  },
  storage: { disks: { default: [Object] } },
  queues: { default: ProxyStub { name: 'WorkerQueue', poisoned: false } },
  channels: {
    default: {
      binding: ProxyStub { name: 'DurableObjectNamespace', poisoned: false }
    }
  }
}

// beginning of getConnection
[class Config] {
  listeners: Map(0) {},
  models: {
    User: [class User extends Model],
    Article: [class Article extends Model]
  },
  events: {
    ArticleUpdated: [class ArticleUpdated extends Event] { shouldQueue: true }
  },
  jobs: { SayHelloJob: [class SayHelloJob extends Job] }
}
```

i’m at a loss as to why the class is different back-to-back like that. i tried logging out the `process.pid` (`import { pid } from "node:process"`) and they are the same. then i thought maybe there could be multiple different versions of superflare installed and that `@superflare/remix` might have a separate instance, but that doesn’t seem to be the case either. i’ve also logged out the point that the `Config` class is declared and i did notice that it gets logged two times with different PIDs at dev startup, then i see one more log on my first request with the same PID as the most recent log before the dev server finishes booting. i’m not sure how that’s possible (that the `Config` class gets declared twice for the same process PID) unless there are in fact multiple installed versions, so i’m gonna try using `console.trace` next to try to get a little more insight. i will update when i have more info, but in case this triggers any thoughts for anyone, i’m documenting it here.

acusti commented:
i’m at a loss as to what’s going on with different `Config` classes being available at different points in the request handling. i temporarily started exporting it from `superflare` so that i can log it out from various places.

when i log the class from within @superflare/remix’s `getLoadContext` util, which is invoked from [remix’s `cloudflare-proxy-plugin.js`](https://github.com/remix-run/remix/blob/main/packages/remix-dev/vite/cloudflare-proxy-plugin.ts), it has all of the expected bindings, but no `models`, `events`, or `jobs`. if i log it from `examples/remix-cms/app/routes/auth/register.tsx`’s `action`, it only has `models`, `events`, and `jobs`.

somehow there are two different `Config` classes that are getting set and imported into various places, but i can’t figure how that could be. i quickly tried seeing what happens if i modify `Config` from a singleton class to a variable on `globalThis`, and that seems fine, so it seems like the runtime context is shared but not the `Config` singleton class. i dunno, i’m struggling to figure out what could be going on. however, the fact that using `globalThis` kinda works suggests that maybe the ASL version @jplhomer you mentioned might actually be a potential fix here. i‘d love to just get it working the way it worked before, but there is something going on with the process lifecycles that i’m missing.

acusti commented:
ok, i realized that the obvious thing that i mentioned yesterday is that the `superflare/src/config.js` file gets executed three times, meaning there must be three separate `Config` classes:

1. the first time is triggered by `esbuild-register` and has a unique PID; i believe this is completely unrelated
2. the second time is after `superflare/cli/dev.ts`’s `devHandler` has run and spawned the `npx remix vite:dev` command, but before the vite server has finished loading. the stack trace points to vite’s `loadConfigFromBundledFile` function, which i am guessing is for loading the `vite-config.ts` file, which does `import { superflareDevProxyVitePlugin } from "@superflare/remix";` and invokes it and passes the result as the first plugin to vite’s `defineConfig`: `superflareDevProxyVitePlugin(config)`
3. the third time is logged when i first hit the server (i.e. visit a route). it has the same PID as the second time, but somehow it seems that another separate `import` is being executed, i guess?

so it seems like this must have something to do with how vite implements import/require. there must be a separate cache or whatever that means the version of superflare imported into our vite plugin is different from the version imported from within the context of the server execution. i’m gonna try using a dynamic import for the vite plugin and see if that allows the import to be handled by vite the same way everything else is handled.

acusti commented:
latest update: the issue was in fact that the vite dev server executes the application, whereas the vite plugin executes outside of the application. this also means that `getLoadContext` is getting invoked from outside the application’s execution context. as a result, i needed to refactor it to take everything that needs to be defined in the application’s execution context as a function argument, so that it can be invoked from the vite plugin but still interop seamlessly with the application itself.

the new implementation cleans up `superflareDevProxyVitePlugin` usage to no longer require any arguments, which is nice, but it does rely on the superflare.config.ts file living at the vite dev server’s root, but i think that might be a reasonable requirement. the plugin now uses `viteDevServer.ssrLoadModule` to load the `superflare` package (for `SuperflareAuth`/`SuperflareSession`) and the user’s `superflare.config.ts` file and provide them all to `getLoadContext`: [f4d25c7](https://github.com/jplhomer/superflare/pull/66/commits/f4d25c7f66aca5dda28651ae7d45cef663cd63f2)

i am still seeing some issues while running the remix-cms example, but i am able to register a user now, so some progress.

jplhomer commented:
@acusti Sounds like you're on the right track. I'm guessing it all boils down to Vite building a separate module graph, and Superflare's config files existing both inside and outside of the Vite process. I've solved it in the past by attaching to something in `globalThis`, but your mileage may vary. I'll try to take a look when I get a free moment soon!

acusti commented:
@jplhomer using `globalThis` would work, but using `ssrLoadModule` also seems to work. my main goal is to avoid major refactoring with these modifications and adaptations.

acusti commented:
@jplhomer i believe the last major missing piece for basic functionality was the logic to `commitSession` if `isDirty()` at the end of every request when running the app using the vite-based dev server. i originally planned to implement that as vite server middleware and register it after the remix `cloudflareDevProxyVitePlugin`. then i discovered that that plugin doesn’t call `next()` on success, which [means that no further middleware](https://github.com/senchalabs/connect?tab=readme-ov-file#use-middleware) will be called after it executes.

so instead, i added the `commitSession` logic to [`entry.server.ts`](https://remix.run/docs/en/main/file-conventions/entry.server). this seems to work, but the downside is that it’s code that must be put into the docs and into the remix template but can’t actually be provided by the @superflare/remix package.

an alternative solution that would mean that would enable us to avoid requiring custom code in the developer’s `entry.server.ts` file would be to re-implement the `cloudflareDevProxyVitePlugin` as a part of the `superflareDevProxyVitePlugin`, so that we can make it call `next()` and then have a final registered middleware that commits the cookie and doesn’t call `next()` to serve as the server terminus. this was actually my original plan until i realized it would mean needing to duplicate all of [this file](https://github.com/remix-run/remix/blob/main/packages/remix-dev/vite/node-adapter.ts) into `@superflare/remix`, which seemed less than ideal, so i didn’t go with that. but it’s totally doable if that seems preferable to you.

the remix-cms example now allows registering and logging in, which is exciting. i need to also update the superflare `dev` CLI command to both run `remix vite:dev` and `wrangler dev` in parallel to [make durable objects work](https://developers.cloudflare.com/workers/wrangler/api/#supported-bindings), but first i’m working on fixing the `AppLoadContext` types.

acusti commented:
@jplhomer i got the `AppLoadContext` types working (it seems to require having a local `./load-context.ts` file to do the type interface re-declaration, but that matches the first-class cloudflare-workers remix template, and the version we need is smaller and simpler), and i fixed the remaining issues with the `remix-cms` example. there are a number of type errors still in that example, but i think most of them are due to the limitations of JSON serializing superflare models (specifically the `Article` model), which i think would be resolved by adopting the [remix single fetch API](https://remix.run/docs/en/main/guides/single-fetch) and its brand of type inference.

i also found out that including the `superflareDevProxyVitePlugin` in `@superflare/remix` means that now, importing anything from `@superflare/remix` results in `wrangler` getting included in whatever bundle is produced, which breaks the build for workers. so, i created a new `@superflare/remix-dev` package and moved the `superflareDevProxyVitePlugin` into that package. now, that package only gets imported into `vite.config.ts`, so none of its dependencies get included in the built worker. maybe there’s another way to make that work, but having a separate package felt like a clean and straightforward way to me.

is it possible to get another `pkg-pr-new` build made from the latest version of this branch so that i can try using it in an external application? i believe all the pieces are now in place.

acusti commented:
@jplhomer thanks for triggering the pkg-pr-new workflow! the build failed because some of the changes in `@superflare/remix` depend on a new type being exported from `superflare`, but running each package build in parallel meant the new type wasn’t available at build time and so typescript reported an error. i removed the `--parallel` flag from the root `build` script, which locally results in the packages being built serially in alphabetical order, so `superflare`, then `superflare-remix`, then `superflare-remix-dev`, which happens to also be the order of the dependency graph, so i think it will work this time.

acusti commented:
@jplhomer one thing i wanted to call out from my previous message: i chose to call the new dev-only package i introduced `@superflare/remix-dev` (to provide `superflareDevProxyVitePlugin`), but i was also considering just calling it `@superflare/dev` (like `@remix-run/dev`). happy to make that change if it sounds better to you. also, i could just keep the `superflareDevProxyVitePlugin` inside the `@superflare/remix` package and have it live in a completely separate file and import path (e.g. `import { superflareDevProxyVitePlugin } from "@superflare/remix/dev";`). i’m also happy to do that if it seems preferable.

acusti commented:
@jplhomer i decided to try implementing a separate entry point in the `@superflare/remix` package for the `superflareDevProxyVitePlugin` (available at `@superflare/remix/dev`). it works well and is a simpler implementation, so there’s no more new superflare package to worry about. however, in order to test it in my own app, i will need another pkg-pr-new workflow to be triggered. sorry for the trouble!

acusti commented:
@jplhomer i used `npx https://pkg.pr.new/jplhomer/superflare@1d12080 new --repo acusti/superflare` (after merging my `wrangler3` branch into `main` in my fork so that it would fetch the updated version of the template) for a new repo i’m working on. i ran into a handful of issues, some with D1 (incorrect config field name + `migrate --fresh` not working), some with type errors due to better worker.ts type coverage and/or the latest workers-types, one with Durable Object config values (all contained in the most recent 11 commits, as of this comment). i am using core pack/yarn berry with PnP in the new repo and using `yarn superflare …` for running commands works quite nicely; i imagine the same will be true with `pnpm superflare …`.

it seems like all of the basics are in place and working. i am aware of two remaining issues with this PR:

- [ ] the ergonomics of running `superflare dev` aren’t great. we have to spawn both `wrangler dev` (if any DOs exist in order to get them to run) and `remix vite:dev`, and the output of both commands shows up in the terminal and makes it unclear what is going on and what URL to visit in order to access the app
- [ ] the documentation needs to be updated based on all of the breaking changes

it would also be worth triggering a new publish-snapshot action so that the latest fixes can be used and properly verified. and please let me know how you think we should proceed, i would love to get this to a place where it actually feels like something that can be merged in and released as the next version of superflare.

dan-gamble commented:
This is fucking sweet @acusti. I've been going off Remix a tad and looking more into Tanstack start but this keeps me super interested in Remix still. The React Router v7 change should be nowhere near as bad as this!

One thing that I don't think will end up related to this PR but likely related to this project at some point in the future is this [Cloudflare Vite Environment repo](https://github.com/flarelabs-net/vite-plugin-cloudflare). I imagine this would help replace the need for the addition of `wrangler dev` alongside `vite:dev`. Just sharing it incase neither of you haven't seen it yet!

acusti commented:
@dan-gamble being able to run Durable Objects from vite would be huge! i wonder what would happen if we added this config to the superflare vite plugin i introduced: https://github.com/flarelabs-net/vite-plugin-cloudflare/blob/main/playground/durable-objects/vite.config.ts

dan-gamble commented:
I think the only issue is not being able to use `@flarelabs-net/vite-plugin-cloudflare` yet. Unless there is a way to get it directly from GitHub or manually copy it all over.

Even just doing the manual copy over would be interesting for sure but I think we'll be waiting on them to get this finalised before being able to properly dip into it.

Hopefully you know more than me and there is a way to properly experiment with it in Superflare 😅

acusti commented:
@jplhomer tl;dr could you please trigger the `publish-snapshot` workflow with the latest so that i can test out those fixes in my own project?

the gory details: i refactored the new `superflareDevProxyVitePlugin` to take over the core part of remix’s `cloudflareDevProxyVitePlugin`, which is the actual [vite dev server middleware](https://github.com/remix-run/remix/blob/main/packages/remix-dev/vite/cloudflare-proxy-plugin.ts#L66-L82) that is used to handle all requests. originally, i [chose not to do so](https://github.com/jplhomer/superflare/pull/66#issuecomment-2285526450) because:

> an alternative solution that would enable us to avoid requiring custom code in the developer’s `entry.server.ts` file would be to re-implement the `cloudflareDevProxyVitePlugin` as a part of the `superflareDevProxyVitePlugin`, so that we can make it call `next()` and then have a final registered middleware that commits the cookie and doesn’t call `next()` to serve as the server terminus. this was actually my original plan until i realized it would require duplicating all of [this file](https://github.com/remix-run/remix/blob/main/packages/remix-dev/vite/node-adapter.ts) into `@superflare/remix`, which seemed less than ideal, so i didn’t go with that. but it’s totally doable if that seems preferable to you.

but a couple of days ago, i noticed that the default `logout` route wasn’t working because it calls `auth.logout()` and returns a `redirect`, meaning that `entry.server.tsx` never gets invoked and so the session never gets committed and the post-logout session data doesn’t get written to the cookie. so it became obvious that taking over the request handling portion of the dev proxy vite plugin was really the best and only way to handle this properly, plus i realized that the `node-adapter.ts` file that i didn’t want to have to duplicate inside superflare is [actually published](https://www.npmjs.com/package/@remix-run/dev?activeTab=code) as `@remix-run/dev/dist/vite/node-adapter.js`, so the surface area of the code that has to be handled by superflare to make the dev server work correctly is actually pretty reasonable, and most of it is code to instantiate `superflare.config`, set up the load context, and then commit the session if `isDirty()` right before generating and streaming the response: https://github.com/jplhomer/superflare/pull/66/files#diff-d9a86b278811d9a84bf4a62d69017ae9f6000d2b9073e8100b7d0ee5e1562041R68-R103

other recent changes are adopting the new [workers Static Assets feature](https://developers.cloudflare.com/workers/static-assets/), upgrading from `turbo` v1 → v2 (latest) and upgrading to the latest version of `pnpm`.

jplhomer commented:
@acusti Just ran again! Thanks for pushing on this - things are crazy for me right now but I anticipate being able to take a look at this and getting it soon. I'd like to get it merged ASAP and switch the project into `prerelease` mode before cutting another major version.

acusti commented:
@jplhomer that’s great news, and thanks for triggering a new release! i will start in on updating the documentation. also, after using the new releases in my own project, i discovered that i wasn’t providing `experimentalJsonConfig: true` by default to wrangler’s `getPlatformProxy`, so i pushed up the [fix for that issue](https://github.com/jplhomer/superflare/pull/66/commits/57eb78ad5bab8ede8df920d3221c38f11c864f31). but aside from that, the new `superflareDevProxyVitePlugin` implementation is working great for me in dev.

jplhomer commented on packages/superflare-remix/dev.ts:
```diff
@@ -0,0 +1,106 @@
+        if (!server.config.server.middlewareMode) {
+          server.middlewares.use(async (nodeReq, nodeRes, next) => {
+            try {
+              // Import singleton-dependent modules via viteDevServer to get
+              // the same instance of the Config singleton class as in app code.
+              const superflare = await server.ssrLoadModule("superflare");
```
Wow, super clever way to share between module graphs.

jplhomer reviewed (approved):
Just phenomenal work. Can't believe you kept up with this. Thank you!

acusti commented:
@jplhomer thanks so much for the kind words! but oh dear, it seems i broke the docs site; let me know if you want any help debugging it. i did have to change a number of things to get it to work with the upgraded stack.

shall i create a new branch to start on documentation updates? there are some subtle changes to document that should be trivial for new superflare consumers but should probably be called out for existing folks looking to upgrade in an upgrade guide or the like. things like with the latest `wrangler`, you have to add `--remote` when running your migrations or else they just get run locally (which i believe wasn’t a thing initially because D1 didn’t even support having a local instance, but i’m not sure). or that if you use `build: { minify: true }` in your vite config, you’ll cause your model names to get mangled and break everything.

jplhomer commented:
@acusti no worries! The only errors I can get from the server are:

<img width="830" alt="Screenshot 2024-11-15 at 2 22 29 PM" src="https://github.com/user-attachments/assets/54521ea0-663c-4d80-9790-72c8b12fa472">

Which seems to correspond to a [Cloudflare 1003](https://developers.cloudflare.com/support/troubleshooting/cloudflare-errors/troubleshooting-cloudflare-1xxx-errors/#error-1003-access-denied-direct-ip-access-not-allowed), something about direct IP references which doesn't add up 🤔 

It also fails locally when I `pnpm start` in the docs site: it's just a Remix 404 page. Very odd - works fine during local development.

I gave it about 30 minutes but can't seem to find any leads - since it's fresher in your mind, maybe you can uncover what might be happening differently in dev vs prod/local build?

jplhomer commented:
Also yes, updating the docs would be great. We haven't added a changeset yet, so there's not a major hurry to get them up.

jplhomer commented:
Update: Found the issue! For some reason, deploying via Cloudflare integration wasn't setting `NODE_ENV=production` like it used to in the GitHub action. Prepending the CF deploy command fixed things 👍

acusti commented:
that’s odd, i haven’t needed to set that manually locally. i figured the most likely thing to break with the docs site was the search integration, but that actually seems to work fine! except for a CSS thing that seems to be the result of the site not having the necessary styles for the `DocSearch-VisuallyHiddenForAccessibility` classname.
